Throw a meanful exception error instead of NPE when no plurals are set

### Is your enhancement related to a problem? Please describe

When you create manually a `ResourceDefinitionContext` and you don't set the plurals field a NPE exception is thrown  when you get it using the Kubernetes client 

```
final ResourceDefinitionContext resourceDefinitionContext = new ResourceDefinitionContext.Builder()
            .withGroup("networking.istio.io")
            .withVersion("v1alpha3")
            .withKind("VirtualService")
            .build();

        System.out.println("Namespace: " + namespace);

        final GenericKubernetesResource virtualResource = kubernetesClient.genericKubernetesResources(resourceDefinitionContext).inNamespace(namespace).withName(virtualServiceName).get();

```

```
Caused by: java.lang.NullPointerException
	at java.base/java.net.URI$Parser.parse(URI.java:3129)
	at java.base/java.net.URI.<init>(URI.java:623)
	at io.fabric8.kubernetes.client.utils.URLUtils.join(URLUtils.java:50)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.getNamespacedUrl(OperationSupport.java:164)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.getNamespacedUrl(OperationSupport.java:169)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getCompleteResourceUrl(BaseOperation.java:818)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getMandatory(BaseOperation.java:188)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:156)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:89)
```

### Describe the solution you'd like

Inform why it is failing.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@
 * Fix #3535: ensure clientKeyAlgo is set properly when loading config YAML from `fromKubeconfig`
 * Fix #3598: applying cancel to the correct future for waitUntilCondition and waitUntilReady
 * Fix #3609: adding locking to prevent long running Watcher methods from causing reconnects with concurrent processing
+* Fix #3620: throw a meaningful exception if no kind/plural is on a ResourceDefinitionContext, default plural if possible
 
 #### Improvements
 

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java
@@ -223,6 +223,7 @@ public class CustomResourceDefinitionContext extends ResourceDefinitionContext {
     }
 
     public CustomResourceDefinitionContext build() {
+      this.customResourceDefinitionContext.validate();
       return this.customResourceDefinitionContext;
     }
   }

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/ResourceDefinitionContext.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/ResourceDefinitionContext.java
@@ -45,6 +45,15 @@ public class ResourceDefinitionContext {
   public boolean isNamespaceScoped() {
     return namespaced;
   }
+  
+  protected void validate() {
+    if (plural == null) {
+      if (kind == null) {
+        throw new IllegalArgumentException("Neither kind nor plural was set, at least one is required");
+      }
+      plural = Utils.getPluralFromKind(kind);
+    }
+  }
 
   public static ResourceDefinitionContext fromResourceType(Class<? extends KubernetesResource> resource) {
     return new Builder()
@@ -89,6 +98,7 @@ public class ResourceDefinitionContext {
     }
     
     public ResourceDefinitionContext build() {
+      this.resourceDefinitionContext.validate();
       return this.resourceDefinitionContext;
     }
   }

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContextTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContextTest.java
@@ -142,8 +142,8 @@ class CustomResourceDefinitionContextTest {
   @Test
   void isNamespaceScoped() {
     // Given
-    CustomResourceDefinitionContext crdc1 = new CustomResourceDefinitionContext.Builder().withScope("Namespaced").build();
-    CustomResourceDefinitionContext crdc2 = new CustomResourceDefinitionContext.Builder().withScope("Cluster").build();
+    CustomResourceDefinitionContext crdc1 = new CustomResourceDefinitionContext.Builder().withPlural("values").withScope("Namespaced").build();
+    CustomResourceDefinitionContext crdc2 = new CustomResourceDefinitionContext.Builder().withPlural("values").withScope("Cluster").build();
 
     // When + Then
     assertThat(crdc1.isNamespaceScoped()).isTrue();

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/base/ResourceDefinitionContextTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/base/ResourceDefinitionContextTest.java
@@ -24,7 +24,7 @@ import static org.junit.jupiter.api.Assertions.assertThrows;
 class ResourceDefinitionContextTest {
 
   @Test
-  void testMissing() {
+  void missingRequiredKindShouldFail() {
     ResourceDefinitionContext.Builder builder = new ResourceDefinitionContext.Builder();
     assertThrows(IllegalArgumentException.class, () -> builder.build());
     builder.withKind("Kind");

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/base/ResourceDefinitionContextTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/base/ResourceDefinitionContextTest.java
@@ -1,0 +1,35 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.fabric8.kubernetes.client.dsl.base;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class ResourceDefinitionContextTest {
+
+  @Test
+  void testMissing() {
+    ResourceDefinitionContext.Builder builder = new ResourceDefinitionContext.Builder();
+    assertThrows(IllegalArgumentException.class, () -> builder.build());
+    builder.withKind("Kind");
+    ResourceDefinitionContext context = builder.build();
+    assertEquals("kinds", context.getPlural());
+  }
+  
+}

